README

INTRODUCTION
------------

The AoEde is the virtual EtherDrive (R) blade, a program that makes a
seekable file available over an ethernet local area network (LAN) via
the ATA over Ethernet (AoE) protocol.

The seekable file is typically a block device like /dev/md0 but even
regular files will work.  Sparse files can be especially convenient.
When AoEde exports the block storage over AoE it becomes a storage
target.  Another host on the same LAN can access the storage if it has
a compatible aoe kernel driver.

BUILDING
--------

The following command should build the AoEde program on a Linux-based
system:

  make

For FreeBSD systems, include an extra parameter like so:

  make PLATFORM=freebsd

For the list of configurable options and their description see config.h

EXAMPLES
--------

There is a "AoEded" script that daemonizes the program and sends its
output to the logger program.  Make sure you have logger installed if
you would like to run AoEde as a daemon with the AoEded script.

  ecashin@kokone AoEde$ echo 'I have logger' | logger
  ecashin@kokone AoEde$ tail -3 /var/log/messages
  Feb  8 14:52:49 kokone -- MARK --
  Feb  8 15:12:49 kokone -- MARK --
  Feb  8 15:19:56 kokone logger: I have logger

Here is a short example showing how to export a block device with a
AoEde.  (This is a loop device backed by a sparse file, but you could
use any seekable file instead of /dev/loop7.)

  ecashin@kokone AoEde$ make
  cc -Wall   -c -o aoe.o aoe.c
  cc -Wall   -c -o linux.o linux.c
  cc -Wall   -c -o ata.o ata.c
  cc -o AoEde aoe.o linux.o ata.o
  ecashin@kokone AoEde$ su
  Password: 
  root@kokone AoEde# modprobe loop
  root@kokone AoEde# dd if=/dev/zero bs=1k count=1 seek=`expr 1024 \* 4096` of=bd
  -file
  1+0 records in
  1+0 records out
  1024 bytes transferred in 0.009901 seconds (103423 bytes/sec)
  root@kokone AoEde# losetup /dev/loop7 bd-file  
  root@kokone AoEde# ./AoEde 9 0 eth0 /dev/loop7 
  ioctl returned 0
  4294968320 bytes
  pid 16967: e9.0, 8388610 sectors

Here's how you can use the Linux aoe driver to access the storage from
another host on the LAN.

  ecashin@kokone ecashin$ ssh makki
  Last login: Mon Feb  7 10:25:04 2005
  ecashin@makki ~$ su
  Password: 
  root@makki ecashin# modprobe aoe
  root@makki ecashin# aoe-stat
      e9.0            eth1              up
  root@makki ecashin# mkfs -t ext3 /dev/etherd/e9.0
  mke2fs 1.35 (28-Feb-2004)
...
  Creating journal (8192 blocks): done
  Writing superblocks and filesystem accounting information: done
  
  This filesystem will be automatically checked every 24 mounts or
  180 days, whichever comes first.  Use tune2fs -c or -i to override.
  root@makki ecashin# mkdir /mnt/e9.0
  root@makki ecashin# mount /dev/etherd/e9.0 /mnt/e9.0
  root@makki ecashin# echo hooray > /mnt/e9.0/test.txt
  root@makki ecashin# cat /mnt/e9.0/test.txt
  hooray

Remember: be as careful with these devices as you would with /dev/hda!

Jumbo Frame Compatibility
-------------------------

AoEde can use jumbo frames provided your initiator is jumbo frame
capable.  There is one small configuration gotcha to consider 
to avoid having the AoEde kernel frequently drop frames.

AoEde uses a raw socket to perform AoE.  The linux kernel will
only buffer a certain amount of data for a raw socket.  For 2.6
kernels, this value is managed through /proc:

root@nai aoe# grep . /proc/sys/net/core/rmem_*
/proc/sys/net/core/rmem_default:128000
/proc/sys/net/core/rmem_max:128000

rmem_max is the max amount a user process may expand the receive
buffer to -- through setsockopt(...) -- and rmem_default is, as you
might expect, the default.

The gotcha is that this amount to buffer does not relate
to the amount of user data buffered, but the amount of
real data buffered.  As an example, the Intel GbE controller
must be given 16KB frames to use an MTU over 8KB.
For each received frame, the kernel must be able to buffer
16KB, even if the aoe frame is only 60 bytes in length.

The linux aoe initiator will use 16 outstanding frames when
used with AoEde.  A good default for ensuring frames are
not dropped is to allocate 16KB for 17 frames:

for f in /proc/sys/net/core/rmem_*; do echo $((17 * 16 * 1024)) >$f; done

Be sure to start AoEde after changing the buffering defaults
as the buffer value is set when the socket is opened.

AoE Initiator Compatibility
---------------------------

The Linux aoe driver for the 2.6 kernel is compatible if you use
aoe-2.6-7 or newer.  You can use older aoe drivers but you will only
be able to see one AoEde per MAC address.
Extended features can be used only with target that supports them.
Currently only AoEDisk does this. You can find it here: 
https://sourceforge.net/projects/aoedisk/


Experimental Tag tracking functionality
---------------------------

Current AoE specification doesn't declare how AoE packet's tag field can
be used by initiator. But at least some of them uses it as monotinically 
incrementing unsigned 32-bit/liddle-endian counter for every new AoE packet. 
This makes possible to reduce duplicated replies transmitted over network 
(RX tags tracking) and to reduce risk of data corruption due to duplicated 
by hardware packet came in wrong order (WRITE tags tracking). 
Both this functionalities are disabled by default and can be enabled by 
specifying command line argument. But be extremely carefull with WRITE tags 
tracking, cause using it with initiator that doesn't meet above behaviour 
can corrupt data! RX tags tracking option is more safe - while used with 
correct initiator it will improve performance, using it with incompatible 
initiator can just cause performance degradation, but not data corruption.
TBD: Add special command in Aoemask request or flag into Aoehdr so initiator 
will explicitely manifest itself as following described tag field behaviour.


CRC data integrity verification 
---------------------------

Normally ethernet hardware uses CRC32 to ensure transferred data integrity 
(google 'Frame Check Sequence' for details). However there're cases when it
would be nice to have some additional integrity verification: to check that
hardware works flawless, or if user is very worried about data integrity and
ready to pay some CPU cycles and/or network throughput. This project is able 
to check CRC of incoming write requests and apply CRC to read responces before
send them. 4 CRC8 checksums used for that. I selected such algorithm since its 
faster than CRC16/32 and still good enough to _diagnose_ errors. However if 
your hardware somewhy doesn't handle own FCS/CRC32 properly - Aoede's CRC is 
not strong enough to guarantee data integrity if used alone.
Note that this feature implemented as extension that should be supported by 
initiator and also must be explicitly activated by it in order to work.
When it activated: any write requests with mismatched CRC will be printed to 
stderr and every read responce will be supplemented with 4 bytes of CRC.


Exported image freeze
---------------------------

AoEde able to freeze disk image for a while with shadowing all writes. This 
means that its possible to stop any modifications to disk image file or device, 
while keeping client to work with it as usual (may be only a bit slower than 
usual). When received 'freeze' signal AoEde flushes all its internal buffers 
and begin functioning in special freezed mode until gets 'unfreeze' signal. 
Also you should specify temporary shadow file, where all writes will go instead 
while actual image remains 'freeze'-d. Use option -f for that, like:
aoede -b 64 1 1 egiga0 /var/.AOE/disk11.fs -f /var/.AOE/shadow11.fs
Also make sure that shadow file located on filesystem that supports sparse files, 
otherwise it will instantly occupy a bit more space than size of main image.
How to freeze/unfreeze? Very simple. Use kill for that:
freeze:
kill -USR1 ${AOEDE_PID}
kill -USR2 ${AOEDE_PID}
unfreeze:
kill -USR1 ${AOEDE_PID}
kill -USR1 ${AOEDE_PID}
kill -USR2 ${AOEDE_PID}
Note that freezing/unfreezing doesn't performed instantly. So after sending freeze 
sequence wait until AoEde will create shadow file. And after sending unfreeze - 
wait until that file will be removed.
Unfreezing flushes written data to main image 'in background', without stopping 
serving initiator requests. However this process noticeable decreases performance 
until finishes. 


Multiple interfaces support
---------------------------

If you want to use AoEde with more than singke network interface you should build it 
with MAX_NICS (in tuneup.h) undefined or set to planned maximum interfaces count.
Then you should launch it specifying  all interfaces is command line, for example:
aoede 1 4 eth0 eth1 /data/AOE.fs
Note that launching several instances of aoede per single file (like it was possible 
with vblade) is not supported due to AoEde has own userspace buffering.



NEWS
-*- change-log -*-
2015-06-14 Ed Cashin <ed.cashin@acm.org>
	Add convenience script for creating sparse files
	vblade-23

2015-02-25 Catalin Salgau <csalgau@users.sourceforge.net>
	Warn about Windows problem with CHS misalignment

2014-09-13 Ed Cashin <ed.cashin@acm.org>
	code cleanup: remove unused variables

2014-08-10 Ed Cashin <ed.cashin@acm.org>
	vblade-22

2014-07-28 Ed Cashin <ed.cashin@acm.org>
	update version for v22 release candidate 1
	buffer boundary cleanups
	FreeBSD BPF and MTU fixes from Catalin Salgau
	offset and size options by Christoph Biedl
	vblade-22rc1

2014-07-27 Killer{R} <support@killprog.com>
	Multiple interfaces support
	Idle timeout now works (periodic buffers flush etc) when compiled without mapped ring buffer support

2014-07-07 Killer{R} <support@killprog.com>
	Avoid memcpy when possible during buffered read (send directly from buffer memory)
	SOCK_RXRING option is off by default (it appeared that stupid but aligned buffer often faster)
	All IO buffers now aligned to page boundary instead of 512 bytes. It also sometimes faster
	Fixed long delay on exit when compiled without SOCK_RXRING
	More flexible selecting blocks/frames count for mapped buffer (if SOCK_RXRING enabled)
	Added additional CRC8x4 data verification support

2014-06-22 Killer{R} <support@killprog.com>
	added Freeze feature
	AIO optimized for FS/device block sizes >512 bytes

2014-06-09 Killer{R} <support@killprog.com>
	added Extensions command to negotiate features 
	removed -t and -T options (negotiated now)
	read requests coalescing (negotiated extension)
	doaoe()&co moved to doaoe.c

2014-06-04 Killer{R} <support@killprog.com>
	added some features and forked as AoEde project:
	using PACKET_RX_RING to receive packets under Linux 
	IO buffering with using of Linux AIO
	tags sequencing to avoid misordered writes 

2013-03-18 Ed Cashin <ecashin@coraid.com>
	add big-endian support from Daniel Mealha Cabrita <dancab@gmx.net>
	vblade-21

2009-08-14 Sam Hopkins <sah@coraid.com>
	bugfix: aoe command error did not set Error bit in flags
	add support for AoEr11
	set ident serial to shelf.slot:hostname
	vblade-20

2008-10-08 Ed Cashin <ecashin@coraid.com>
	add Chris Webb's bpf fix for FreeBSD
	add Ryan Thomas's fix to stop bufcnt being overridden
	vblade-19
	
2008-07-14 Ed Cashin <ecashin@coraid.com>
	add Chris Webb's block device options patch
	add Chris Webb's socket options patch for better jumbo handling
	remove obsolete contrib/o_direct.diff
	vblade-18

2008-06-09 Ed Cashin <ecashin@coraid.com>
	add Chris Webb's latest BPF patch to vblade, remove from contrib
	update contributed AIO patch for compatibility with current sources
	vblade-17
	
2008-05-07 Ed Cashin <ecashin@coraid.com>
	add Chris Webb's AIO patch to the contributions
	add Chris Webb's BPF patch to the contributions
	vblade-16

2008-02-20 Ed Cashin <ecashin@coraid.com>
	require the amount of data we use, not the amount ethernet requires
	make sure the packet length agrees with the config query length
	make sure the packet length agrees with the amount to write
	remove newline embedded in fw version field of ATA dev ID response
	vblade-15
	
2006-11-20 Sam Hopkins <sah@coraid.com>
	apply contrib jumbo patch to standard distribution
	add jumbo configuration app. note in README
	add jumbo README reference to manpage
	add mask feature; -m flag
	update manpage to describe -m flag
	vblade-14

2006-10-05 Sam Hopkins <sah@coraid.com>
	fix confcmd memcpy bug
	correct scnt return value in read/write ata response
	replace O_RDONLY fallback with explicit stat.  root always wins.
	vblade-13

2006-10-04 Sam Hopkins <sah@coraid.com>
	fix confcmd buglets
	fix atacmd buglets
	add atacmd handling for bad argument errors
	add O_RDONLY open if O_RDWR fails
	add contrib patch directory
	add contrib/README
	add jumbo patch to contrib
	add o_direct patch to contrib
	vblade-12

2006-09-21 "Adam J. Richter" <adam@yggdrasil.com>
	add install target for makefile
	vblade-11
	
2005-12-06 Ed Cashin <ecashin@coraid.com>
	fix u64 configuration on FreeBSD
	release vblade-10
	
2005-12-06 Valeriy Glushkov <valery@rocketdivision.com>
	implemented config string support
	added handler for ATA Check power mode command
	
2005-11-15 Ed Cashin <ecashin@coraid.com>
	add compatibility with platforms lacking u64 (e.g., Slackware)
	release vblade-9
	
2005-11-10 Ed Cashin <ecashin@coraid.com>
	call atainit on program startup
	put VBLADE_VERSION in dat.h and use it in firmware version
	release vblade-7
	include Stacey's patch to use p{read,write} on FreeBSD

aoe.c
/*
  Copyright 2005-2014, CORAID.
  For contact information, see http://coraid.com/
  Copyright 2014, Killer{R}
  For contact information, see http://killprog.com/

  This file is part of AoEde.

  AoEde is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 2 of the License.

  AoEde is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with AoEde.  If not, see <http://www.gnu.org/licenses/>.
*/

#define _GNU_SOURCE
#include "config.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <malloc.h> 
#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <sys/file.h>
#include <endian.h>
#include <fcntl.h>
#include <netinet/in.h>
#include "dat.h"
#include "fns.h"
#include <netinet/if_ether.h>

enum {
	Nmasks= 32,
	Nsrr= 256,
	Alen= 6,
};

uchar masks[Nmasks*Alen];
int nmasks;
uchar srr[Nsrr*Alen];
int nsrr;
char config[Nconfig];
int nconfig = 0;
int maxscnt = 2;
char *ifname;
int bufcnt = Bufcount;

static void
aoead()	// advertise the virtual blade
{
	update_maxscnt();	
	Conf *p = (Conf *)alloca(2000);
	int i;
	memset(p, 0, sizeof *p);
	memset(p->h.dst, 0xff, 6);
	memcpy(p->h.src, nics[curnic].mac, 6);
	p->h.type = type_net;
	p->h.flags = Resp;
	p->h.maj = shelf_net;
	p->h.min = slot;
	p->h.cmd = Config;
	p->bufcnt = htons(bufcnt);
	p->scnt = nics[curnic].maxscnt;
	p->firmware = htons(FWV);
	p->vercmd = 0x10 | Qread;
	memcpy(p->data, config, nconfig);
	p->len = htons(nconfig);
	if (nmasks == 0)
	if (putpkt((uchar *)p, sizeof *p - sizeof p->data + nconfig) == -1) {
		perror("putpkt aoe id");
	}

	for (i=0; i<nmasks; i++) {
		memcpy(p->h.dst, &masks[i*Alen], Alen);
		if (putpkt((uchar *)p, sizeof *p - sizeof p->data + nconfig) == -1)
			perror("putpkt aoe id");
	}
}

int
isbcast(uchar *ea)
{
	uchar *b = (uchar *)"\377\377\377\377\377\377";

	return memcmp(ea, b, 6) == 0;
}

long long
getlba(uchar *p)
{
	vlong v;
	int i;

	v = 0;
	for (i = 0; i < 6; i++)
		v |= (vlong)(*p++) << i * 8;
	return v;
}

int
aoeata(Ata *p, int pktlen)	// do ATA reqeust
{
	Ataregs r;
	int len = 60;
	int n;

	r.lba = getlba(p->lba);
	r.sectors = p->sectors;
	r.feature = p->err;
	r.cmd = p->cmd;
	if (r.cmd != 0xec)
	if (!rrok(p->h.src)) {
		p->h.flags |= Error;
		p->h.error = Res;
		return len;
	}
	if (atacmd(&r, (uchar *)(p+1), maxscnt*512, pktlen - sizeof(*p)) < 0) {
		p->h.flags |= Error;
		p->h.error = BadArg;
		return len;
	}
	if (!(p->aflag & Write))
	if ((n = p->sectors)) {
		n -= r.sectors;
		len = sizeof (Ata) + (n*512);
	}
	p->sectors = r.sectors;
	p->err = r.err;
	p->cmd = r.status;
	return len;
}

#define QCMD(x) ((x)->vercmd & 0xf)

// yes, this makes unnecessary copies.

int
confcmd(Conf *p, int payload)	// process conf request
{
	int len;

	len = ntohs(p->len);
	if (QCMD(p) != Qread)
	if (len > Nconfig || len > payload)
		return 0;	// if you can't play nice ...
	switch (QCMD(p)) {
	case Qtest:
		if (len != nconfig)
			return 0;
		// fall thru
	case Qprefix:
		if (len > nconfig)
			return 0;
		if (memcmp(config, p->data, len))
			return 0;
		// fall thru
	case Qread:
		break;
	case Qset:
		if (nconfig)
		if (nconfig != len || memcmp(config, p->data, len)) {
			p->h.flags |= Error;
			p->h.error = ConfigErr;
			break;
		}
		// fall thru
	case Qfset:
		nconfig = len;
		memcpy(config, p->data, nconfig);
		break;
	default:
		p->h.flags |= Error;
		p->h.error = BadArg;
	}
	memmove(p->data, config, nconfig);
	p->len = htons(nconfig);
	p->bufcnt = htons(bufcnt);
	p->scnt = maxscnt = (getmtu(sfd, ifname) - sizeof (Ata)) / 512;
	p->firmware = htons(FWV);
	p->vercmd = 0x10 | QCMD(p);	// aoe v.1
	return nconfig + sizeof *p - sizeof p->data;
}

static int
aoesrr(Aoesrr *sh, int len)
{
	uchar *m, *e;
	int n;

	e = (uchar *) sh + len;
	m = (uchar *) sh + Nsrrhdr;
	switch (sh->rcmd) {
	default:
e:		sh->h.error = BadArg;
		sh->h.flags |= Error;
		break;
	case 1:	// set
		if (!rrok(sh->h.src)) {
			sh->h.error = Res;
			sh->h.flags |= Error;
			break;
		}
	case 2:	// force set
		n = sh->nmacs * 6;
		if (e < m + n)
			goto e;
		nsrr = sh->nmacs;
		memmove(srr, m, n);
	case 0:	// read
		break;
	}
	sh->nmacs = nsrr;
	n = nsrr * 6;
	memmove(m, srr, n);
	return Nsrrhdr + n;
}

static int
addmask(uchar *ea)
{

	uchar *p, *e;

	p = masks;
	e = p + nmasks;
	for (; p<e; p += 6)
		if (!memcmp(p, ea, 6))
			return 2;
	if (nmasks >= Nmasks)
		return 0;
	memmove(p, ea, 6);
	nmasks++;
	return 1;
}

static void
rmmask(uchar *ea)
{
	uchar *p, *e;

	p = masks;
	e = p + nmasks;
	for (; p<e; p+=6)
		if (!memcmp(p, ea, 6)) {
			memmove(p, p+6, e-p-6);
			nmasks--;
			return;
		}
}

static int
aoemask(Aoemask *mh, int len)
{
	Mdir *md, *mdi, *mde;
	int i, n;

	n = 0;
	md = mdi = (Mdir *) ((uchar *)mh + Nmaskhdr);
	switch (mh->cmd) {
	case Medit:
		mde = md + mh->nmacs;
		for (; md<mde; md++) {
			switch (md->cmd) {
			case MDdel:
				rmmask(md->mac);
				continue;
			case MDadd:
				if (addmask(md->mac))
					continue;
				mh->merror = MEfull;
				mh->nmacs = md - mdi;
				goto e;
			case MDnop:
				continue;
			default:
				mh->merror = MEbaddir;
				mh->nmacs = md - mdi;
				goto e;
			}
		}
		// success.  fall thru to return list
	case Mread:
		md = mdi;
		for (i=0; i<nmasks; i++) {
			md->res = md->cmd = 0;
			memmove(md->mac, &masks[i*6], 6);
			md++;
		}
		mh->merror = 0;
		mh->nmacs = nmasks;
		n = sizeof *md * nmasks;
		break;
	default:
		mh->h.flags |= Error;
		mh->h.error = BadArg;
	}
e:	return n + Nmaskhdr;
}

void
doaoe(Aoehdr *p, int n)
{
	int len;

	switch (p->cmd) {
	case ATAcmd:
		if (n < Natahdr)
			return;
		len = aoeata((Ata*)p, n);
		break;
	case Config:
		if (n < Ncfghdr)
			return;
		len = confcmd((Conf *)p, n);
		break;
	case Mask:
		if (n < Nmaskhdr)
			return;
		len = aoemask((Aoemask *)p, n);
		break;
	case Resrel:
		if (n < Nsrrhdr)
			return;
		len = aoesrr((Aoesrr *)p, n);
		break;
	default:
		p->error = BadCmd;
		p->flags |= Error;
		len = n;
		break;
	}
	if (len <= 0)
		return;
	memmove(p->dst, p->src, 6);
	memmove(p->src, mac, 6);
	p->maj = htons(shelf);
	p->min = slot;
	p->flags |= Resp;
	if (putpkt(sfd, (uchar *) p, len) == -1) {
		perror("write to network");
		exit(1);
	}
}

void
aoe(void)
{
	Aoehdr *p;
	uchar *buf;
	int n, sh;
	long pagesz;
	enum { bufsz = 1<<16, };

	if ((pagesz = sysconf(_SC_PAGESIZE)) < 0) {
		perror("sysconf");
		exit(1);
	}        
	if ((buf = malloc(bufsz + pagesz)) == NULL) {
		perror("malloc");
		exit(1);
	}
	n = (size_t) buf + sizeof(Ata);
	if (n & (pagesz - 1))
		buf += pagesz - (n & (pagesz - 1));

	aoead(sfd);

	for (;;) {
		n = getpkt(sfd, buf, bufsz);
		if (n < 0) {
			perror("read network");
			exit(1);
		}
		if (n < sizeof(Aoehdr))
			continue;
		p = (Aoehdr *) buf;
		if (ntohs(p->type) != 0x88a2)
			continue;
		if (p->flags & Resp)
			continue;
		sh = ntohs(p->maj);
		if (sh != shelf && sh != (ushort)~0)
			continue;
		if (p->min != slot && p->min != (uchar)~0)
			continue;
		if (nmasks && !maskok(p->src))
			continue;
		doaoe(p, n);
	}
}

void
usage(void)
{
	fprintf(stderr, "usage: %s [-b bufcnt] [-o offset] [-l length] [-d ] [-s] [-r] [ -m mac[,mac...] ] shelf slot netif filename\n", 
		progname);
	exit(1);
}

/* parseether from plan 9 */
int
static void 
aoeready()
{
	int n, prevcurnic = curnic;
	for (n = 0; n<niccnt; ++n) {
		curnic = n;
		aoead();
	}
	curnic = prevcurnic;
	malloc_trim(0);
}

static void *
allocate_ata_aligned_buffer(size_t sz)
{
	//kernel 2.6 requires 512 bytes alignment for O_DIRECT
	//however direct I/O works faster with page-aligned buffers
	void *buf;
	if ((buf = malloc(sz + page_size)) == NULL) {
		perror("malloc");
		grace_exit(1);
	}
	if ((sz = ((size_t) buf + sizeof(Ata)) % page_size)!=0)
		buf+= (page_size - sz);

	return buf;
}

static void 
aoe(void) 
{
	//jumbo frames nowadays are 9K maximum
	//32K ought to be enough for anybody ;)
	const size_t bufsz = 0x8000 - page_size;
	int n, m;
	void *buf, *buf_pkt_in;

#ifdef KEEP_STATS
	time_t tm_stats = time(0); 
#endif

	buf =  allocate_ata_aligned_buffer(bufsz);
#ifdef SOCK_RXRING
	if (rxring_init()!=-1) {
		aoeready();
		rxring_roll(buf);	
		rxring_deinit();
		grace_exit(1);
	}
	printf("Falling back to socket read\n");
#endif

	buf_pkt_in = allocate_ata_aligned_buffer(bufsz);
	aoeready();
	for (;;) {
		Aoehdr *p;
		n = getpkt(buf_pkt_in, bufsz);
		if (AOE_UNLIKELY(n < 0)) {
			perror("read network");
			grace_exit(1);
		}
		p = (Aoehdr *) buf_pkt_in;
		m = packet_check(p, n);
		if (AOE_LIKELY(m>=0)) {
			if (tags_tracking==TAGS_INC_LE || tags_tracking==TAGS_INC_BE) 
				tagring_select(m);
			doaoe(p, (Aoehdr *)buf, n);
		}

#ifdef KEEP_STATS
		if (tags_tracking!=TAGS_ANY && skipped_writes) {
			time_t tm_now = time(0); 
			if ((tm_stats+60)<tm_now || tm_now<tm_stats) {
				printf("skipped_writes=%llu\n", skipped_writes);
				tm_stats = tm_now;
			}
		}
#endif
	}
}

static void
print_usage_and_die(void)
{
	fprintf(stderr, "AoEde project. (http://aoede.sourceforge.net/)\n");
	fprintf(stderr, "This application licensed under the GPLv2 (http://www.gnu.org/licenses/gpl-2.0.html)\n");
	fprintf(stderr, "This work based on vblade that is created by CORAID (www.coraid.com)\n");

	fprintf(stderr, "usage: %s [-b bufcnt] [-f path[,size]] [-d ] [-s] [-r] [-t] [-T] [ -m mac[,mac...] ] shelf slot netif1 [netif2 [..]]] filename\n", 
		progname);
	fprintf(stderr, "options:\n");
	fprintf(stderr, " -b  specify socket x-fer buffers in MTU units (linux only)\n");
	fprintf(stderr, " -f  specify freeze storage with optional size limit\n");
	fprintf(stderr, " -r  give only read access to disk image\n");
	fprintf(stderr, " -m  restrict accessing disk only from specified mac address(es)\n");
	
	grace_exit(1);
}

/* parseether from plan 9 */
static int
parseether(uchar *to, char *from)
{
	char nip[4];
	char *p;
	int i;

	p = from;
	for(i = 0; i < 6; i++){
		if(*p == 0)
			return -1;
		nip[0] = *p++;
		if(*p == 0)
			return -1;
		nip[1] = *p++;
		nip[2] = 0;
		to[i] = strtoul(nip, 0, 16);
		if(*p == ':')
			p++;
	}
	return 0;
}

static void
setmask(char *ml)
{
	char *p;
	int n;

	for (; ml; ml=p) {
		p = strchr(ml, ',');
		if (p)
			*p++ = '\0';
		n = parseether(&masks[nmasks*Alen], ml);
		if (n < 0)
			fprintf(stderr, "ignoring mask %s, parseether failure\n", ml);
		else
			nmasks++;
	}
}

int
maskok(uchar *ea)
{
	int i, ok = 0;

	for (i=0; !ok && i<nmasks; i++)
		ok = memcmp(ea, &masks[i*Alen], Alen) == 0;
	return ok;
}

static void
setserial(char *f, int sh, int sl)
{
	char h[32];

	h[0] = 0;
	gethostname(h, sizeof h);
	snprintf(serial, Nserial, "%d.%d:%.*s", sh, sl, (int) sizeof h, h);
}

int
maskcheck(uchar *ea)
{	
	int i;
	if (!nmasks)
		return 0;

	for (i=0; i<nmasks; ++i)
	{
		if (memcmp(ea, &masks[i*Alen], Alen) == 0)
			return i;
	}
	return -1;
}

int
rrok(uchar *ea)
{
	int i, ok = 0;

	if (nsrr == 0)
		return 1;
	for (i=0; !ok && i<nsrr; i++)
		ok = memcmp(ea, &srr[i*Alen], Alen) == 0;
	return ok;
}

void
setserial(int sh, int sl)
{
	char h[32];

	h[0] = 0;
	gethostname(h, sizeof h);
	snprintf(serial, Nserial, "%d.%d:%.*s", sh, sl, (int) sizeof h, h);
}

static void
open_bfd_or_die(char *filename, int omode)
{
	struct stat st;
	uchar readonly = ((omode&O_RDWR)!=O_RDWR);
	bfd = open(filename, omode);
	if (bfd == -1) {
		perror("open");
		grace_exit(1);
	}
	if (!readonly && flock(bfd, LOCK_EX|LOCK_NB)<0) {
		perror("flock");
		grace_exit(1);
	}
	if (fstat(bfd, &st)==0) {
		if (st.st_blksize > SECTOR_SIZE && 
			st.st_blksize < (0x100*SECTOR_SIZE) && 
			(st.st_blksize % SECTOR_SIZE)==0) {
			bfd_blocks_per_sector = (uchar) (st.st_blksize / SECTOR_SIZE);
		}
	}

	size = getsize(bfd);
	size /= SECTOR_SIZE;
	if (size==0) {
		fprintf(stderr, "Empty exported file/device.\n");
		grace_exit(1);
	}

	printf("pid %ld: e%d.%d, %lld sectors%s%s%s, block/sector: %u\n",
		(long) getpid(), shelf, slot, size,
		readonly ? " O_RDONLY" : " O_RDWR",  

#ifdef O_DIRECT
		(omode&O_DIRECT) ? " O_DIRECT" : "",
#else
		"",
#endif

#ifdef O_DSYNC 
		(omode&O_DSYNC) ? " O_DSYNC" : "",
#else
		(omode&O_SYNC) ? " O_SYNC" : "",
#endif

		bfd_blocks_per_sector);
}

static void
open_nics_or_die(char **names)
{
#ifndef MAX_NICS
	nics = (struct NIC *)malloc(niccnt * sizeof(struct NIC));
	if (!nics) {
		perror("malloc");
		grace_exit(1);
	}
	memset(nics, 0, niccnt * sizeof(struct NIC));
#endif

	for (curnic = 0; curnic<niccnt; ++names, ++curnic) {
		printf("NIC %s: ", *names);
		if ( (nics[curnic].sfd = dial(*names, bufcnt))<0) {
			perror("dial");
			grace_exit(1);
		}
		if (getea(nics[curnic].sfd, *names, nics[curnic].mac)<0) {
			perror("getea");
			grace_exit(1);
		}
		nics[curnic].name = *names;
		update_maxscnt();
		printf( "mac=%02X:%02X:%02X:%02X:%02X:%02X maxscnt=%u\n", 
			nics[curnic].mac[0], nics[curnic].mac[1], nics[curnic].mac[2], 
			nics[curnic].mac[3], nics[curnic].mac[4], nics[curnic].mac[5], nics[curnic].maxscnt );
	}
	curnic = 0;
}

int
main(int argc, char **argv)
{
	char *sz;
#ifdef KEEP_STATS
	skipped_writes = 0;
	skipped_packets = 0;
#endif
	int ch, omode = O_NOATIME | O_RDWR;
	page_size = getpagesize();
#ifdef USE_AIO
# ifdef O_DIRECT
	omode |= O_DIRECT;
# else
#  ifdef O_DSYNC 
	omode |= O_DSYNC;
#  else
	omode |= O_SYNC;
#  endif
# endif
#endif

	bufcnt = Bufcount;
	setbuf(stdin, NULL);
	progname = *argv;
	tags_tracking = TAGS_ANY;
	while ((ch = getopt(argc, argv, "f:b:dsrm:")) != -1) {
		switch (ch) {
		case 'b':
			bufcnt = atoi(optarg);
			break;

		case 'f':
#ifdef SHADOW_FREEZE
			freeeze_path = strdup(optarg);
			sz = strrchr(freeeze_path, ',');
			if (sz) {
				*sz = 0;
				freeeze_size_limit = atoi(sz+1);
				printf("Freeze storage '%s' with %lluMB size limit\n", freeeze_path, freeeze_size_limit);
				freeeze_size_limit*= 0x100000;
			}
			else
				printf("Freeze storage '%s' without size limit\n", freeeze_path);
#else
			printf("Freeze storage can't be used due to SHADOW_FREEZE not specified in build\n");
#endif
			break;

		case 'd':
#ifdef O_DIRECT
# ifdef USE_AIO
			printf("Direct mode implied by AIO\n");
# else
			omode |= O_DIRECT;
# endif
#else
			printf("Direct IO is not supported in this build\n");
#endif
			break;
		case 's':
# ifdef O_DSYNC 
			omode |= O_DSYNC;
# else
			omode |= O_SYNC;
# endif
			break;
		case 'r':
			omode &= ~(int)O_RDWR;
			omode |= O_RDONLY;
			break;
		case 'm':
			setmask(optarg);
			break;
		case 'o':
			offset = strtoll(optarg, &end, 0);
			if (end == optarg || offset < 0)
				usage();
			break;
		case 'l':
			length = strtoll(optarg, &end, 0);
			if (end == optarg || length < 1)
				usage();
			break;
		case '?':
		default:
			print_usage_and_die();
		}
	}
	argc -= optind;
	argv += optind;
	if (argc < 4 || bufcnt <= 0)
		print_usage_and_die();
	shelf = atoi(argv[0]);
	shelf_net = htons(shelf);
	type_net = htons(0x88a2);
	slot = atoi(argv[1]);
	setserial(argv[3], shelf, slot);
#ifdef MAX_NICS
	if ((argc - 3)>MAX_NICS) {
		printf("This build supports maximum %u NICs\n", MAX_NICS);
		grace_exit(1);
	}
#endif

#if !defined(MAX_NICS) || (MAX_NICS>1)
	niccnt = argc - 3;
#endif

	open_bfd_or_die(argv[argc - 1], omode);
	open_nics_or_die(argv + 2);
	iox_init(); 
	tagring_init();
	atainit();
	fflush(stdout);
	aoe();
	return 0;
}

void 
grace_exit(int ncode) //flush IO buffers and exit
{
	iox_flush();
	exit(ncode);
}
